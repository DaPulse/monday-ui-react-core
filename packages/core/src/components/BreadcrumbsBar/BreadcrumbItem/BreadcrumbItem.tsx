import { ComponentDefaultTestId, getTestId } from "../../../tests/test-ids-utils";
import cx from "classnames";
import React, { useRef } from "react";
import useIsOverflowing from "../../../hooks/useIsOverflowing/useIsOverflowing";
import Tooltip from "../../../components/Tooltip/Tooltip";
import { BreadcrumbContent } from "./BreadcrumbContent/BreadcrumbContent";
import { HideShowEvent } from "../../../constants";
import { SubIcon, VibeComponentProps } from "../../../types";
import styles from "./BreadcrumbItem.module.scss";

const MOUSEENTER = [HideShowEvent.MOUSE_ENTER];
const MOUSELEAVE = [HideShowEvent.MOUSE_LEAVE];

export interface BreadcrumbItemProps extends VibeComponentProps {
  /** The display text. */
  text?: string;
  /** Should item be disabled. */
  disabled?: boolean;
  /** Should item be clickable - this should be recieved from the breadcrumbsBar ( Navigation/Indication bar ). */
  isClickable?: boolean;
  /** If the item is clickable and the type of navigation is a link, this is the link */
  link?: string;
  /** If the item is clickable and the type of navigation is a function, this is the function */
  onClick?: () => void;
  /** Should be the current Item - mainly effects the item`s style. */
  isCurrent?: boolean;
  /** An Icon - If no icon needed then should be left empty. */
  icon?: SubIcon;
}

const BreadcrumbItem: React.FC<BreadcrumbItemProps> = ({
  className,
  text = "",
  disabled,
  isClickable = false,
  link,
  onClick,
  isCurrent = false,
  icon,
  id,
  "data-testid": dataTestId
}) => {
  const componentRef = useRef<HTMLSpanElement>(null);
  const isOverflowing = useIsOverflowing({ ref: componentRef });

  return (
    <Tooltip
      disableDialogSlide={true}
      withoutDialog={false}
      content={isOverflowing && text}
      showTrigger={MOUSEENTER}
      hideTrigger={MOUSELEAVE}
    >
      <li
        id={id}
        data-testid={dataTestId || getTestId(ComponentDefaultTestId.BREADCRUMB_ITEM, id)}
        className={cx(styles.breadcrumbItemWrapper, className, {
          [styles.disabled]: disabled
        })}
      >
        <BreadcrumbContent
          ref={componentRef}
          isClickable={isClickable}
          link={link}
          onClick={onClick}
          text={text}
          icon={icon}
          isCurrent={isCurrent}
          disabled={disabled}
        />
      </li>
    </Tooltip>
  );
};

export default BreadcrumbItem;
