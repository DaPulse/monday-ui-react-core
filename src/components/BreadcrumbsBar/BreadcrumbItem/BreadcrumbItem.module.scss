.breadcrumbItemWrapper {
  display: inline-flex;
  overflow: hidden;
  list-style-type: none;
  cursor: default;
}

.breadcrumbItemWrapper.clickable {
  cursor: pointer;
}

.breadcrumbItemWrapper.disabled {
  cursor: not-allowed;
}
