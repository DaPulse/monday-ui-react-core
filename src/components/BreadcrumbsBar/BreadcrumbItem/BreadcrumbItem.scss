@import "../../../styles/themes.scss";
@import "../../../styles/states.scss";

.breadcrumbItem--wrapper {
  display: inline-flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-between;
  font-size: 14px;
  @include theme-prop(color, secondary-text-color);
  padding: 0px 2px;
  overflow: hidden;  

  .breadcrumb-text {
    color: inherit;
    text-decoration: none;
    margin: 0px 4px;
    line-height: 22px;

    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
  }

  .breadcrumb-icon {
    height: 14px;
    width: 14px;
    flex-shrink: 0;
  }

  &.hover:hover {
    @include theme-prop(background-color, primary-background-hover-color);
    border-radius: 4px;
  }

  &.disabled {
    @include disabled;
  }

  &.current {
    @include theme-prop(color, primary-text-color);
    font-weight: 500;
  }
}
