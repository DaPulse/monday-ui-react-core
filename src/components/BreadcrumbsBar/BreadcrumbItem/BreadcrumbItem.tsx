import { ELEMENT_TYPES, getTestId } from "../../../utils/test-utils";
import cx from "classnames";
import React, { useRef } from "react";
import useIsOverflowing from "../../../hooks/useIsOverflowing";
import Tooltip from "../../../components/Tooltip/Tooltip";
import { backwardCompatibilityForProperties } from "../../../helpers/backwardCompatibilityForProperties";
import { BreadcrumbContent } from "./BreadcrumbContent/BreadcrumbContent";
import VibeComponentProps from "../../../types/VibeComponentProps";
import { IconSubComponentProps } from "../../Icon/Icon";
import { HideShowEvent } from "../../../constants";
import styles from "./BreadcrumbItem.module.scss";

const MOUSEENTER = [HideShowEvent.MOUSE_ENTER];
const MOUSELEAVE = [HideShowEvent.MOUSE_LEAVE];

export interface BreadcrumbItemProps extends VibeComponentProps {
  /** The display text. */
  text?: string;
  /** Should item be disabled. */
  disabled?: boolean;
  /** Backward compatibility for props naming */
  isDisabled?: boolean;
  /** Should item be clickable - this should be recieved from the breadcrumbsBar ( Navigation/Indication bar ). */
  isClickable?: boolean;
  /** If the item is clickable and the type of navigation is a link, this is the link */
  link?: string;
  /** If the item is clickable and the type of navigation is a function, this is the function */
  onClick?: () => void;
  /** Should be the current Item - mainly effects the item`s style. */
  isCurrent?: boolean;
  /** An Icon - If no icon needed then should be left empty. */
  // TODO TS-migration fix icon type?
  icon?: string | React.FunctionComponent<IconSubComponentProps> | null;
}

const BreadcrumbItem: React.FC<BreadcrumbItemProps> = ({
  className,
  text = "",
  disabled,
  // Backward compatibility for props naming
  isDisabled,
  isClickable = false,
  link,
  onClick,
  isCurrent = false,
  icon,
  id,
  "data-testid": dataTestId
}) => {
  const overrideDisabled = backwardCompatibilityForProperties([disabled, isDisabled], false) as boolean;
  const componentRef = useRef<HTMLSpanElement>(null);
  const isOverflowing = useIsOverflowing({ ref: componentRef });

  return (
    <Tooltip
      disableDialogSlide={true}
      withoutDialog={false}
      content={isOverflowing && text}
      showTrigger={MOUSEENTER}
      hideTrigger={MOUSELEAVE}
    >
      <li
        id={id}
        data-testid={dataTestId || getTestId(ELEMENT_TYPES.BREADCRUMB_ITEM, id)}
        className={cx(styles.wrapper, "breadcrumbItem--wrapper", className, {
          [styles.clickable]: isClickable,
          ["clickable"]: isClickable,
          [styles.disabled]: overrideDisabled,
          ["disabled"]: overrideDisabled
        })}
      >
        <BreadcrumbContent
          ref={componentRef}
          isClickable={isClickable}
          link={link}
          onClick={onClick}
          text={text}
          icon={icon}
          isCurrent={isCurrent}
          disabled={overrideDisabled}
        />
      </li>
    </Tooltip>
  );
};

export default BreadcrumbItem;
