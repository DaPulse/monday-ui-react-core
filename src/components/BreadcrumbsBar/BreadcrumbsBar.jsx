import React, { useState, useRef, useEffect } from "react";
import PropTypes from "prop-types";
import cx from "classnames";
import NavigationChevronRight from "../Icon/Icons/components/NavigationChevronRight";
import "./BreadcrumbsBar.scss";

const BreadcrumbsBar = ({ className, children, hasHoverEffect = true }) => {
  const [activeItemIndex, setActiveItemIndex] = useState(-1);
  const componentRef = useRef(null);

  useEffect(()=>{
    console.log(componentRef)
    //componentRef && componentRef.current && componentRef.current.focus();
  }, [componentRef])

  return (  
    <nav aria-label="Breadcrumb" className={cx("breadcrumbs-bar--wrapper", className)}>
      <ol>
        {children &&
          React.Children.map(children, (child, index) => [
            React.cloneElement(child, {
              ...child?.props,
              isActive: index == activeItemIndex,
              index,
              setActiveItemIndex,
              hasHover: hasHoverEffect
            }),
            index < children.length - 1 && (
              <NavigationChevronRight className="sparatorIcon" size={"14"} aria-hidden="true" />
            )
          ])}
      </ol>
    </nav>
  );
};

BreadcrumbsBar.propTypes = {
  className: PropTypes.string,
  hasHoverEffect: PropTypes.bool
};
BreadcrumbsBar.defaultProps = {
  className: "",
  hasHoverEffect: true
};

export default BreadcrumbsBar;
