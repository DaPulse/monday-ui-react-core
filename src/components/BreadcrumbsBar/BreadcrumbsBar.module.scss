@import "../../styles/themes";

.breadcrumbsBarWrapper ol {
  display: flex;
  flex-direction: row;
  align-items: center;
  white-space: nowrap;
}

.breadcrumbsBarWrapper ol .separatorIcon {
  @include theme-prop(color, secondary-text-color);
  flex-shrink: 0;
}
