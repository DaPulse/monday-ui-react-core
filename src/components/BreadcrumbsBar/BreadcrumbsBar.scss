@import "../../styles/themes.scss";

.breadcrumbs-bar--wrapper ol{
    display: flex;
    flex-direction: row;
    align-items: center;
    white-space: nowrap;

    .sparatorIcon{
        @include theme-prop(color, secondary-text-color);
        flex-shrink: 0;
    }
}
