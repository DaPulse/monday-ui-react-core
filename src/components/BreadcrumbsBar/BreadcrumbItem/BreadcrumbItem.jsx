import React, { useRef } from "react";
import PropTypes from "prop-types";
import classNames from "classnames";
import useIsOverflowing from "../../../hooks/useIsOverflowing";
import Tooltip from "../../Tooltip/Tooltip";
import { BreadcrumbContent } from "./BreadcrumbContent/BreadcrumbContent";
import "./BreadcrumbItem.scss";

const MOUSEENTER = ["mouseenter"];
const MOUSELEAVE = ["mouseleave"];

const BreadcrumbItem = ({ className, text, isDisabled, isClickable, link, onClick, isCurrent, icon }) => {
  const componentRef = useRef(null);
  const isOverflowing = useIsOverflowing({ ref: componentRef });

  return (
    <Tooltip
      disableDialogSlide={true}
      withoutDialog={false}
      content={isOverflowing && text}
      showTrigger={MOUSEENTER}
      hideTrigger={MOUSELEAVE}
    >
      <li
        className={classNames(
          "breadcrumbItem--wrapper",
          className,
          { clickable: isClickable },
          { current: isCurrent },
          { disabled: isDisabled }
        )}
      >
        <BreadcrumbContent
          className={classNames(
            "breadcrumb-content",
            { clickable: isClickable },
            { current: isCurrent },
            { disabled: isDisabled }
          )}
          ref={componentRef}
          isClickable={isClickable}
          link={link}
          onClick={onClick}
          text={text}
          icon={icon}
          isCurrent={isCurrent}
        />
      </li>
    </Tooltip>
  );
};

BreadcrumbItem.propTypes = {
  className: PropTypes.string,
  /** The display text. */
  text: PropTypes.string,
  /** Should item be disabled. */
  isDisabled: PropTypes.bool,
  /** Should item be clickable - this should be recieved from the breadcrumbsBar ( Navigation/Indication bar ). */
  isClickable: PropTypes.bool,
  /** If the item is clickable and the type of navigation is a link, this is the link */
  link: PropTypes.string,
  /** If the item is clickable and the type of navigation is a function, this is the function */
  onClick: PropTypes.func,
  /** Should be the current Item - mainly effects the item`s style. */
  isCurrent: PropTypes.bool,
  /** An Icon - If no icon needed then should be left empty. */
  icon: PropTypes.string
};

BreadcrumbItem.defaultProps = {
  className: "",
  text: "",
  isDisabled: false,
  isClickable: false,
  link: undefined,
  onClick: undefined,
  isCurrent: false,
  icon: undefined
};

export default BreadcrumbItem;
