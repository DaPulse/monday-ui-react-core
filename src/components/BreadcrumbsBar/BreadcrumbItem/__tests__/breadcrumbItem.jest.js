import React from "react";
import renderer from "react-test-renderer";
import BreadcrumbsBar from "../../BreadcrumbsBar";
import BreadcrumbItem from "../BreadcrumbItem";
import BoardIcon from "../../../Icon/Icons/components/Board";
import WorkspaceIcon from "../../../Icon/Icons/components/Workspace";
import { fireEvent, render } from "@testing-library/react";

jest.useFakeTimers();

describe("BreadcrumbsItem", () => {
  it("renders correctly with empty props", () => {
    const tree = renderer
      .create(
        <BreadcrumbsBar>
          <BreadcrumbItem />
        </BreadcrumbsBar>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders correctly with icon", () => {
    const tree = renderer
      .create(
        <BreadcrumbsBar>
          <BreadcrumbItem
            icon={WorkspaceIcon}
            text="Workspace"
          />
          <BreadcrumbItem
            icon={BoardIcon}
            text="Board"
          />
        </BreadcrumbsBar>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("if navigation item, href link is correct", () => {
    const { getByRole } = render(
      <BreadcrumbsBar type={BreadcrumbsBar.types.NAVIGATION}>
        <BreadcrumbItem text="Workspace" link="https://www.google.com" />
      </BreadcrumbsBar>
    );

    const link = getByRole("link");

    expect(link.getAttribute('href')).toBe("https://www.google.com");
  });

  it("if navigation item, click works", () => {
    const onClickMock = jest.fn();

    const { getByText } = render(
      <BreadcrumbsBar type={BreadcrumbsBar.types.NAVIGATION}>
        <BreadcrumbItem text="Workspace" onClick={onClickMock} />
      </BreadcrumbsBar>
    );

    const item = getByText("Workspace");
    fireEvent.click(item);
    jest.advanceTimersByTime(1000);

    expect(onClickMock.mock.calls.length).toBe(1);
  });

  it("if indication item, click does nothing", () => {
    const onClickMock = jest.fn();

    const { getByText } = render(
      <BreadcrumbsBar type={BreadcrumbsBar.types.INDICATION}>
        <BreadcrumbItem text="Workspace" onClick={onClickMock} />
      </BreadcrumbsBar>
    );

    const item = getByText("Workspace");
    fireEvent.click(item);
    jest.advanceTimersByTime(1000);

    expect(onClickMock.mock.calls.length).toBe(0);
  });

  it("click with Enter key works", () => {
    const onClickMock = jest.fn();

    const { getByText } = render(
      <BreadcrumbsBar type={BreadcrumbsBar.types.NAVIGATION}>
        <BreadcrumbItem text="Workspace" onClick={onClickMock} />
      </BreadcrumbsBar>
    );

    const item = getByText("Workspace");
    fireEvent.focus(item);
    fireEvent.keyDown(item, { key: "Enter", code: "Enter", charCode:"13" });
    jest.advanceTimersByTime(1000);

    expect(onClickMock.mock.calls.length).toBe(1);
  });

  it("click with Space key works", () => {
    const onClickMock = jest.fn();

    const { getByText } = render(
      <BreadcrumbsBar type={BreadcrumbsBar.types.NAVIGATION}>
        <BreadcrumbItem text="Workspace" onClick={onClickMock} />
      </BreadcrumbsBar>
    );

    const item = getByText("Workspace");
    fireEvent.keyDown(item, { key: " ", code: "Space", charCode:"32" });
    jest.advanceTimersByTime(1000);

    expect(onClickMock.mock.calls.length).toBe(1);
  });

  it("'current' item renders correctly", () => {
    const tree = renderer
      .create(
        <BreadcrumbsBar>
          <BreadcrumbItem isCurrent={true} />
        </BreadcrumbsBar>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("'disabled' item renders correctly", () => {
    const tree = renderer
      .create(
        <BreadcrumbsBar>
          <BreadcrumbItem isDisabled={true} />
        </BreadcrumbsBar>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
