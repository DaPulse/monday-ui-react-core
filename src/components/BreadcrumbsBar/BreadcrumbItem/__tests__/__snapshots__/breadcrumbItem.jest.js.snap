// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BreadcrumbsItem 'current' item renders correctly 1`] = `
<nav
  aria-label="Breadcrumb"
  className="breadcrumbs-bar--wrapper"
>
  <ol>
    <li
      className="breadcrumbItem--wrapper current"
    >
      <span
        aria-current="page"
        aria-disabled="true"
        className="breadcrumb-content current"
        tabIndex="0"
      >
        <span
          className="breadcrumb-text"
        >
          
        </span>
      </span>
    </li>
  </ol>
</nav>
`;

exports[`BreadcrumbsItem 'disabled' item renders correctly 1`] = `
<nav
  aria-label="Breadcrumb"
  className="breadcrumbs-bar--wrapper"
>
  <ol>
    <li
      className="breadcrumbItem--wrapper disabled"
    >
      <span
        aria-disabled="true"
        className="breadcrumb-content disabled"
        tabIndex="0"
      >
        <span
          className="breadcrumb-text"
        >
          
        </span>
      </span>
    </li>
  </ol>
</nav>
`;

exports[`BreadcrumbsItem renders correctly with empty props 1`] = `
<nav
  aria-label="Breadcrumb"
  className="breadcrumbs-bar--wrapper"
>
  <ol>
    <li
      className="breadcrumbItem--wrapper"
    >
      <span
        aria-disabled="true"
        className="breadcrumb-content"
        tabIndex="0"
      >
        <span
          className="breadcrumb-text"
        >
          
        </span>
      </span>
    </li>
  </ol>
</nav>
`;

exports[`BreadcrumbsItem renders correctly with icon 1`] = `
<nav
  aria-label="Breadcrumb"
  className="breadcrumbs-bar--wrapper"
>
  <ol>
    <li
      className="breadcrumbItem--wrapper"
    >
      <span
        aria-disabled="true"
        className="breadcrumb-content"
        tabIndex="0"
      >
        <svg
          className="breadcrumb-icon"
          clickable={false}
          fill="currentColor"
          height="14"
          viewBox="0 0 20 20"
          width="14"
        >
          <path
            clipRule="evenodd"
            d="M3 2.25C2.58579 2.25 2.25 2.58579 2.25 3V8.38462C2.25 8.79883 2.58579 9.13462 3 9.13462H8.38462C8.79883 9.13462 9.13462 8.79883 9.13462 8.38462V3C9.13462 2.58579 8.79883 2.25 8.38462 2.25H3ZM3.75 7.63462V3.75H7.63462V7.63462H3.75ZM11.6154 2.25C11.2012 2.25 10.8654 2.58579 10.8654 3V8.38462C10.8654 8.79883 11.2012 9.13462 11.6154 9.13462H17C17.4142 9.13462 17.75 8.79883 17.75 8.38462V3C17.75 2.58579 17.4142 2.25 17 2.25H11.6154ZM12.3654 7.63462V3.75H16.25V7.63462H12.3654ZM2.25 11.6154C2.25 11.2012 2.58579 10.8654 3 10.8654H8.38462C8.79883 10.8654 9.13462 11.2012 9.13462 11.6154V17C9.13462 17.4142 8.79883 17.75 8.38462 17.75H3C2.58579 17.75 2.25 17.4142 2.25 17V11.6154ZM3.75 12.3654V16.25H7.63462V12.3654H3.75ZM11.6154 10.8654C11.2012 10.8654 10.8654 11.2012 10.8654 11.6154V17C10.8654 17.4142 11.2012 17.75 11.6154 17.75H17C17.4142 17.75 17.75 17.4142 17.75 17V11.6154C17.75 11.2012 17.4142 10.8654 17 10.8654H11.6154ZM12.3654 16.25V12.3654H16.25V16.25H12.3654Z"
            fill="currentColor"
            fillRule="evenodd"
          />
        </svg>
        <span
          className="breadcrumb-text"
        >
          Workspace
        </span>
      </span>
    </li>
    <svg
      aria-hidden="true"
      className="sparatorIcon"
      fill="currentColor"
      height="14"
      viewBox="0 0 20 20"
      width="14"
    >
      <path
        clipRule="evenodd"
        d="M13.5303 9.46967L13 10L13.5303 10.5303C13.8232 10.2374 13.8232 9.76256 13.5303 9.46967ZM11.9393 10L5.46967 16.4697C5.17678 16.7626 5.17678 17.2374 5.46967 17.5303C5.76256 17.8232 6.23744 17.8232 6.53033 17.5303L13.5303 10.5303L13 10L13.5303 9.46967L6.53033 2.46967C6.23744 2.17678 5.76256 2.17678 5.46967 2.46967C5.17678 2.76256 5.17678 3.23744 5.46967 3.53033L11.9393 10Z"
        fill="currentColor"
        fillRule="evenodd"
      />
    </svg>
    <li
      className="breadcrumbItem--wrapper"
    >
      <span
        aria-disabled="true"
        className="breadcrumb-content"
        tabIndex="0"
      >
        <svg
          className="breadcrumb-icon"
          clickable={false}
          fill="currentColor"
          height="14"
          viewBox="0 0 20 20"
          width="14"
        >
          <path
            clipRule="evenodd"
            d="M7.5 4.5H16C16.2761 4.5 16.5 4.72386 16.5 5V15C16.5 15.2761 16.2761 15.5 16 15.5H7.5L7.5 4.5ZM6 4.5H4C3.72386 4.5 3.5 4.72386 3.5 5V15C3.5 15.2761 3.72386 15.5 4 15.5H6L6 4.5ZM2 5C2 3.89543 2.89543 3 4 3H16C17.1046 3 18 3.89543 18 5V15C18 16.1046 17.1046 17 16 17H4C2.89543 17 2 16.1046 2 15V5Z"
            fill="currentColor"
            fillRule="evenodd"
          />
        </svg>
        <span
          className="breadcrumb-text"
        >
          Board
        </span>
      </span>
    </li>
  </ol>
</nav>
`;
